src/tls/keypair.h

```diff
@@ -501,43 +470,128 @@ namespace tls
       SECP256K1_CONTEXT_VERIFY | SECP256K1_CONTEXT_SIGN);
-    secp256k1_pubkey c4_pub;
-#endif
+
+    static constexpr size_t privk_size = 32;
+    uint8_t c4_priv[privk_size] = {0};
 
   public:
-    /**
-     * Construct from a public key in PEM format
-     *
-     * @param public_pem Sequence of bytes containing the key in PEM format
-     */
-    PublicKey(const std::vector<uint8_t>& public_pem)
+    template <typename... Ts>
+    KeyPair_k1Bitcoin(Ts... ts) : KeyPair(std::forward<Ts>(ts)...)
     {
-      mbedtls_pk_init(&ctx);
-      mbedtls_pk_parse_public_key(&ctx, public_pem.data(), public_pem.size());
+      if (
+        mbedtls_mpi_write_binary(
+          &(mbedtls_pk_ec(*key)->d), c4_priv, privk_size) != 0)
+        throw std::logic_error("Could not extract raw private key");
 
-#if CURVE_CHOICE_SECP256K1_BITCOIN
-      auto k = mbedtls_pk_ec(ctx);
-      size_t pub_len;
-      uint8_t pub_buf[100];
-      int rc = mbedtls_ecp_point_write_binary(
-        &k->grp, &k->Q, MBEDTLS_ECP_PF_COMPRESSED, &pub_len, pub_buf, 100);
-      if (rc)
-        throw std::logic_error("mbedtls_ecp_point_write_binary failed");
-      rc = secp256k1_ec_pubkey_parse(ctx_, &c4_pub, pub_buf, pub_len);
-      if (rc != 1)
-        throw std::logic_error("ecp256k1_ec_pubkey_parse failed");
-#endif
+      if (secp256k1_ec_seckey_verify(k1_ctx, c4_priv) != 1)
+        throw std::logic_error("secp256k1 private key is not valid");
+    }
+
+    ~KeyPair_k1Bitcoin()
+    {
+      if (k1_ctx)
+        secp256k1_context_destroy(k1_ctx);
+    }
+
+    int sign_hash(
+      const uint8_t* hash,
+      size_t hash_size,
+      size_t* sig_size,
+      uint8_t* sig) const override
+    {
+      if (hash_size != 32)
+        return -1;
+
+      secp256k1_ecdsa_signature k1_sig;
+      if (
+        secp256k1_ecdsa_sign(
+          k1_ctx, &k1_sig, hash, c4_priv, nullptr, nullptr) != 1)
+        return -2;
+
+      if (
+        secp256k1_ecdsa_signature_serialize_der(
+          k1_ctx, sig, sig_size, &k1_sig) != 1)
+        return -3;
+
+      return 0;
+    }
+  };
+
+  using KeyPairPtr = std::shared_ptr<KeyPair>;
+
+  /**
+   * Create a new public / private key pair on specified curve and
+   * implementation
+   */
+  template <typename... Ts>
+  inline KeyPairPtr make_key_pair(
+    CurveImpl curve = CurveImpl::ledger_curve_choice)
+  {
+    const auto ec = get_ec_for_curve_impl(curve);
+
+    if (curve == CurveImpl::secp256k1_bitcoin)
+    {
+      return KeyPairPtr(new KeyPair_k1Bitcoin(ec));
+    }
+    else
+    {
+      return KeyPairPtr(new KeyPair(ec));
+    }
+  }
+
+  /**
+   * Create a public / private from existing raw private key data
+   */
+  template <typename... Ts>
+  inline KeyPairPtr make_key_pair(
+    CBuffer pkey,
+    CBuffer pw = nullb,
+    bool use_bitcoin_impl = prefer_bitcoin_secp256k1)
+  {
+    std::unique_ptr<mbedtls_pk_context> key =
+      std::make_unique<mbedtls_pk_context>();
+    mbedtls_pk_init(key.get());
```

Comment:
can this fail?

Reply:
[Documentation ](https://tls.mbed.org/api/pk_8h.html#a999d1160bb30c03d0c4382c3a9b0aa89) doesn't suggest so. Return type is void.